vmware-tanzu/nsx-operator
# Support NSX default project for SecurityPolicy

Starting from VPC 2.0, it's not allowed to created objects under /orgs/default/projects/default/infra path. So for NetworkPolicy/SecurityPolicy with namespaceSelector, in order to create Groups under Default Project, it's needed to create them under /infra/domains/default/groups/<>. As for Groups under non default Project, it's still allowed to create them under /orgs/default/projects/\<custom project\>/infra/domains/default/groups.

This patch is to:
1. Support NetworkPolicy/SecurityPolicy creation under Default Project.
2. Refactor VPC SecurityPolicy HAPI call process for both creation and deletion.
3. Refactor VPC SecurityPolicy store apply process after creation and deletion.


Test done:
 1. Create SecurityPolicy and Networkpolicy with namespaceSelector in VPC mode for default project and non-default project, and delete them successfully.
 2. Create SecurityPolicy in T1 mode, and delete them successfully.
 3. Create SecurityPolicy and Networkpolicy with namespaceSelector in VPC mode, and delete them from K8s firstly, and do GC to see created NSX SecurityPolicies are deleted by GC.
 4. Create SecurityPolicy in T1 mode, and delete them from K8s firstly, and do GC to see created NSX SecurityPolicies are deleted by GC.
 5. Create  SecurityPolicy with namespaceSelector, hack the code to do an error test:
-  Make operator reboot right after InfraClient.Patch to create infra shares/groups successfully. 
-  By that time, SP K8s CR is created with error status. NSX infra shares/groups created, but NSX SecurityPolicy is not created. 
-  Reboot operator and make code go the normal process, the NSX SecurityPolicy  is created successfully, and SP K8s CR is back to normal status. 

 6.  Delete  SecurityPolicy with namespaceSelector, hack the code to do an error test:
- Make operator reboot right after OrgRootClient.Patch to delete SecurityPolicy from NSX.
- By that time, NSX infra shares/groups are not deleted, K8s SP CR is not deleted.
- Reboot operator and make code go the normal process,  K8s SP CR is deleted since there is no the corresponding NSX SecurityPolicy in the store, but we still have infra shares/groups in the store.
- GC process will delete stale infra shares/groups.
